extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -0,0 +1,1212 @@
+    \[
+      \textrm{For all } i \in t.\inputs,\enspace i.\validator^{\#} = \txout(i, \lambda).\addr.
+    \]
+  \end{enumerate}
+  \caption{$\sigma$-validity of a transaction $t$ in the XUTXO-2 model}
+  \label{fig:xutxo-2-validity}
+
+\end{ruledfigure}
+
+We use the definitions of \unspent{}  and
+\getvalue{} from Section~\ref{sec:xutxo-1-validity} unchanged,
+but we also require a function to retrieve the values of
+custom currencies from inputs. If $t$ is a transaction in a ledger
+$\lambda$ and $i \in t.\inputs$, then we define 
+$$
+\s{getCustomValues}(i,\lambda) = \lambda\langle i.\outputref.\txid \rangle.\outputs[i.\outputref.\idx].\customvals
+$$
+\noindent If any of the indexing operations here fail then the
+transaction is invalid.
+
+\paragraph{Note.} In rule~\ref{rule:custom-values-are-preserved-2}, $+$ and $\sum$ are the
+sum of maps as defined in Figure~\ref{fig:basic-notation}. Essentially
+we require that the quantities of each of the individual custom currencies
+involved in the transaction are preserved.  Recall that values in
+$\customforge$ can be negative whereas values in outputs must
+be non-negative.  Thus rule~\ref{rule:custom-values-are-preserved-2} implies
+that a transaction is invalid if it attempts to destroy more of a currency
+than is actually available in its inputs.
+
+
+\paragraph{Monetary policies for custom currencies.}
+The new \textbf{Forging}
+rule enables custom currencies to implement their own monetary
+policies: for example, one might wish to place some limit on the
+amount of a currency that can be forged, or restrict the creation of
+the currency to owners of particular public keys.
+
+The idea is that a custom currency has a monetary policy which is
+defined by some script $H$, and the hash $h = H^{\#}$ is used as the
+identifier of the currency\footnotemark.  Whenever a new quantity of
+the currency is forged, rules~\ref{rule:custom-forge},
+\ref{rule:all-inputs-validate-2}, and
+\ref{rule:validator-scripts-hash-2} imply that $H$ must be executed;
+$H$ is provided with the \forge{} field of the transaction via
+$\sigma$, and so it knows how much of the currency is to be forged and
+can respond appropriately.
+
+The advantage of this scheme is that custom currencies can be handled
+entirely within the smart contract system, without the need to
+introduce any extra blockchain infrastructure such as a central
+registry of custom currencies.
+
+\footnotetext{In fact it is not quite enough to use the hash of the
+  script here, because many different currencies might wish to use the
+  same policy and hence the same script.  One way to remedy this is to
+  embed a reference to an unspent output in the script and say that
+  the currency can only be forged if the referenced output is spent at
+  the same time, so it can only be forged once. \red{I don't think
+    that's quite what we want here though: we have one opportunity
+    to forge the currency, then we can't make any more.}}
```

Comment:
I think there are two different concerns:

1. Using the same monetary policy multiple times
2. Preventing unauthorised forging of a currency

The "referencing an unspent output" trick accomplishes both but is mainly aimed at (2). You could also achieve (1) by embedding any kind of random data in the validator script, like a nonce.